### DISCUSSION TITLE
TSwap approve function

### DISCUSSION THREAD

**Discussion by Shawon-Mandal** [Cyfrin/security-and-auditing-full-course-s23#168] [Q&A] [answered]
I am currently writing the TSwap invariant tests along with the course. 
In the `setUP` function of `invariant.t.sol`  there are these lines

``` javascript 

        poolToken.approve(address(pool), type(uint256).max);
        weth.approve(address(pool), type(uint256).max);
        
 ```
I could not find any approve function in `TSwapPool.sol` but this `approve` was
 found after running  `forge inspect TSwapPool methods` command  in the terminal.

In short, if there is no approve function in the original `TSwapPool.sol` is it
 mandatory to approve them inside the invariant test suite? if yes, then why?
 

Thank you

**Comment by EngrPips** [comment, marked as answer]
Hello @Shawon-Mandal, I am not sure I correctly understand your question, but we are doing those two approvals because we want to send some `poolToken` and `weth` to the pool. Since both `poolToken`and `weth` are `ERC20,` we need to call the approve function on them to approve the address of the `TSWAP pool` to take our token. Doing so ensures that the transaction successfully goes through whenever we call a function to supply liquidity to the `TSWAP pool`

**Reply by Shawon-Mandal** [reply]
I see. I was confused at first to see that when I wrote the `forge inspect TSwapPool methods` inside my terminal, I saw this output 

![approve](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/99707719/c4def507-1eef-42c7-9a64-dc01ab2bf5ca)

I thought that the approve function belonged to TSwapPool.sol contract due to the reason that TSwapPool was mentioned in the command. Now I realize that it was actually a function of ERC20.sol

**Reply by EngrPips** [reply]
I am glad you realize it now. Keep crushing it, friend.
